src/audio/kpb.c

```diff
@@ -171,45 +179,56 @@ static int kpb_prepare(struct comp_dev *dev)
 
 	cd->no_of_clients = 0;
 
-	/* allocate history_buffer/s */
-#if KPB_NO_OF_HISTORY_BUFFERS == 2
+	/* allocate history buffer/s */
+	while (hb_size > 0 && i < ARRAY_SIZE(hb_mcp)) {
+		ptr = rballoc(RZONE_RUNTIME, hb_mcp[i], ca_size);
+
+		if (ptr) {
+			history_buffer->start_addr = ptr;
+			history_buffer->end_addr = ptr + ca_size;
+			history_buffer->w_ptr = ptr;
+			history_buffer->r_ptr = ptr;
+			history_buffer->state = KPB_BUFFER_FREE;
+			hb_size = hb_size - ca_size;
+			history_buffer->next = &cd->history_buffer;
+
+			/* Do we need another buffer? */
+			if (hb_size > 0) {
+				new_hb = rzalloc(RZONE_RUNTIME,
+						 SOF_MEM_CAPS_RAM,
+						 COMP_SIZE(struct hb));
+				history_buffer->next = new_hb;
+				new_hb->state = KPB_BUFFER_OFF;
+				new_hb->next = history_buffer;
+				history_buffer = new_hb;
```

Comment:
this isnt clear either why would history_buffer->next and history_buffer both point to new_hb?

Reply:
ahh I see what you are doing now. But is it supposed to be a circular linked list of history buffers?

Reply:
Yes, correct.